firedrake/ensemble/ensemble_functionspace.py

```diff
@@ -0,0 +1,242 @@
+        """
+        return self._mesh
+
+    @cached_property
+    def dual(self):
+        """The dual to this function space.
+        A :class:`firedrake.EnsembleDualSpace` if self is a :class:`firedrake.EnsembleFunctionSpace`, and vice-versa.
+        """
+        return EnsembleFunctionSpace(
+            [V.dual() for V in self.local_spaces], self.ensemble)
+
+    @cached_property
+    def nlocal_spaces(self):
+        """The total number of subspaces across all ensemble ranks.
+        """
+        return len(self.local_spaces)
+
+    @cached_property
+    def nglobal_spaces(self):
+        """The total number of subspaces across all ensemble ranks.
+        """
+        return self.ensemble_comm.allreduce(len(self.local_spaces))
+
+    @property
+    def nlocal_rank_dofs(self):
+        """The total number of dofs across all subspaces on the local MPI rank.
+        """
+        return self._full_local_space.node_set.size
+
+    @property
+    def nlocal_comm_dofs(self):
+        """The total number of dofs across all subspaces on the local ensemble.comm.
+        """
+        return self._full_local_space.dim()
+
+    @property
+    def nglobal_dofs(self):
+        """The total number of dofs across all subspaces on all ensemble ranks.
+        """
+        return self.ensemble_comm.allreduce(self.nlocal_rank_dofs)
+
+    def _component_indices(self, i):
+        """
+        Return the indices into the local mixed function storage
+        corresponding to the i-th local function space.
+        """
+        offset = sum(len(V) for V in self.local_spaces[:i])
+        return tuple(offset + i for i in range(len(self.local_spaces[i])))
+
+    def create_vec(self):
+        """Return a PETSc Vec on the global_comm with the same layout as a :class:`firedrake.EnsembleFunction` or :class:`firedrake.EnsembleCofunction` in this function space.
+        """
+        vec = PETSc.Vec().create(comm=self.global_comm)
+        vec.setSizes((self.nlocal_dofs, self.nglobal_dofs))
+        vec.setUp()
+        return vec
+
+    def local_to_global_ises(self, i):
+        # COMM_SELF or global_comm?
+        pass
+
+    def __eq__(self, other):
+        if not isinstance(other, type(self)):
+            local_eq = False
+        elif other.ensemble is not self.ensemble:
+            # TODO: should we relax this to allow congruent ensembles?
+            local_eq = False
+        elif len(other.subfunctions) != len(self.subfunctions):
+            local_eq = False
+        else:
+            local_eq = all(
+                lspace == rspace
+                for lspace, rspace in zip(self.local_spaces,
+                                          other.local_spaces)
+            )
+        return all(self.ensemble.ensemble_comm.allgather(local_eq))
+
+    def __neq__(self, other):
+        return not self.__eq__(other)
+
+
+class EnsembleFunctionSpace(EnsembleFunctionSpaceBase):
+    def __new__(cls, local_spaces, ensemble):
```

Comment:
As above, I don't love this

Reply:
Resolving, see comments on consistency above.